### PR TITLE
doc: Add software architecture diagram and design considerations

### PR DESCRIPTION
**Pull request by elsalahy**
#### Summary

Added SW architecture diagram, with explanations and design considerations.

#### Screenshots

Screen Shot 2021-10-08 at 13 58 32

#### Changes

- Add SW architecture diagram
- Add explanations to the different SW layers

#### Notes for Reviewers

...

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Run Locally: Verified that the docs build using `make server`, posted screenshots, verified external links.
- [x] New Features Marked: Documentation for new features is marked using the `new-in-version` shortcode, according to the guidelines in [CONTRIBUTING](CONTRIBUTING.md).
- [x] Style Guidelines: Documentation obeys style guidelines in [CONTRIBUTING](CONTRIBUTING.md).
- [x] Commits: Commit messages follow guidelines in [CONTRIBUTING](CONTRIBUTING.md), there are no fixup commits left.

### CODE REVIEW

**Comment by elsalahy** [comment]
Closes #43

File: doc/content/sensor-edition/software/se-sw-arch/_index.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
+---
+title: "Sensor Edition Software Architecture"
+description: ""
+weight: -1
+distributions: null
+---
+
+The {{% gnse %}} software contains multiple applications, libraries and drivers.
+
+The {{% gnse %}} software architecture observed below depicts three main layers
```
Comment (nejraselimovic) [inline doc/content/sensor-edition/software/se-sw-arch/_index.md:10]:
```suggestion
The {{% gnse %}} software architecture observed below depicts three main layers:
```

File: doc/content/sensor-edition/software/se-sw-arch/_index.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+### Library
+
+The library (lib) layer contains functions and APIs that abstract the hardware and target layers, thus allowing users to build applications that are independent of the used hardware components.
+
+Most of the libraries are built by third-parties such as the hardware manufacturer, and all of them are integrated within the {{% gnse %}} platform to ensure a better experience when building applications.
+
+Different applications use different libraries depending on the application desired functionality.
+
+For example, an application that samples sensors and sends the data using LoRaWAN, will utilize the Sensors, GNSE APIs, bare-metal scheduler and LoRaWAN libraries.
```
Comment (nejraselimovic) [inline doc/content/sensor-edition/software/se-sw-arch/_index.md:37]:
```suggestion
For example, an application that samples sensors and sends the data using LoRaWAN will utilize the Sensors, GNSE APIs, bare-metal scheduler and LoRaWAN libraries.
```

File: doc/content/sensor-edition/software/se-sw-arch/_index.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+If the developer needs to build an application that uses the external flash memory or the secure element, they can rely on the Storage or Security libraries.
+
+## Application
+
+The application layer contains sample boilerplate applications that can be used as a starting point.
+The example applications are developed to be as energy efficient as possible and can be used for production testing and deployments.
```
Comment (nejraselimovic) [inline doc/content/sensor-edition/software/se-sw-arch/_index.md:44]:
```suggestion
The application layer contains sample boilerplate applications that can be used as a starting point.

The example applications are developed to be as energy efficient as possible and can be used for production testing and deployments.
```

File: doc/content/sensor-edition/software/se-sw-arch/_index.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+Examples in the application layer should invoke the library layer functions/APIs and shouldn't directly invoke target layer functionalities unless that is the only way to achieve the desired functionality.
```
Comment (nejraselimovic) [inline doc/content/sensor-edition/software/se-sw-arch/_index.md:46]:
```suggestion
Examples in the application layer should invoke the library layer functions/APIs and should not directly invoke target layer functionalities unless that is the only way to achieve the desired functionality.
```

**Review by nejraselimovic — APPROVED** [review]

**Comment by elsalahy** [comment]
Thanks @nejraselimovic for the helpful feedback, can you please have another final review, thanks!

**Pull request merged**